hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BalancerChore.java

```diff
@@ -46,7 +52,20 @@ public BalancerChore(HMaster master) {
   @Override
   protected void chore() {
     try {
-      master.balance();
+      BalanceResponse balanceResponse = master.balance();
+      //TODO: Now, only StochasticLoadBalancer has implemented the method
+      // 'updateBalancerLoadInfo'. Code here should be more generic if other balancers implement
+      // the method 'updateBalancerLoadInfo'
+      if (!balanceResponse.isBalancerRan()) {
```

Comment:
So the key change is here. When balancer is not run, we will get the current assignment snapshot and update the load info?

Reply:
This covers one of the possible holes, but not all of them. For example, if you look at the implementation of `StochasticLoadBalancer#balanceTable`, there is a check for `!needsBalance(tableName, cluster)`. If that's true, we return null early. In that case, isBalancerRan would be true, but metrics would not have been updated.

One way around this would be to also check `balanceResponse.getMovesCalculated() == 0` here. However, I think that is too much indirection. Instead I would recommend you re-organize balanceTable a bit so that we still compute costs and call `updateStochasticCosts` even if `!needsBalance`. I think this would be a matter of moving needsBalance check down below the curOverallCost and tempFunctionCosts handling, and add a call to updateStochasticCosts there as well.

One other change I might suggest making is at the end of balanceTable, we should move the updateStochasticCosts call into the `if (initCost > currentCost)` branch. At this point, the costs have been re-computed based on the moves attempted and it's very possible that the moves chosen actually worsened the costs, but those moves were not accepted. So we should only update costs based on currentcost if we actually accepted the change.

Reply:
Thanks for your suggestion. What about we remove `if (!balanceResponse.isBalancerRan())` and just as you said "call this method at the start of balanceTable and this would unify with the normal balance call path"，we call 'updateBalancerLoadInfo' directly regardless of  the balance result.

This will not change any balance logic of StochasticLoadBalancer.

Reply:
I think I might prefer keeping the isBalancerRan check you have here, and make the change in balanceTable like I suggested. I think it's ok to change the balancer logic there. Here's my reasons, and let me know if you disagree:

- The balancer already calls updateStochasticCosts twice
- The second call to updateStochasticCosts (at the end of balanceTable) is wrong -- it should happen in the `if (initCost > currentCost)` branch. This is a bug we should fix.
- Adding an unchecked call to updateBalancerLoadInfo here would mean that now we're calling updateStochasticCosts three times

So the normal case balancer chore run would:

- call updateBalancerLoadInfo, which computes costs and updatesStochasticCosts
- call balancer, which will also computeCosts and call updateStochasticCosts twice

(or the other way around).

I think it'd be more intuitive and less work to keep this chore how you have it, and update balanceTable so that if the balancer actually runs, it ensures the correct costs metrics are reported even if it doesn't calculate any moves. This change here is just ensuring that we _also_ update cost metrics if the balancer _doesn't_ run.

Does that make sense? My feedback about unifying the updateBalancerLoadInfo/computeCosts implementations would remain.

Reply:
Perfect! I got your point and aggree with you.

Reply:
I had one more thought -- now that we're using computeCosts directly (good), it along with initCosts, etc operates directly on balancer instance variables. This is usually protected by synchronization in the upstream caller, notably in HMaster.balance.

I think we should try to similarly synchronize calling updateBalancerLoadInfo. One thought I had was whether you could move the below logic into a new method on HMaster, similar to the balance() method. So here your code would be:

```java
BalanceResponse response = master.balance()
if (!response.isBalancerRan()) {
  master.updateBalancerMetrics();
}
```

updateBalancerMetrics would encapsulate the below code within a `synchronized (this.balancer)`.

Alternatively, we could move this whole chore into an HMaster.balanceOrUpdateMetrics() and call that here.

Thoughts?

Reply:
I have declared the updateBalancerLoadInfo in RSGroupBasedLoadBalancer as synchronized in the previous commit(Just like "updateClusterMetrics" is also synchronized), so maybe we don’t need `synchronized(this.balancer)`?  I am not really sure it is ok, let me know if there is any problem.

I like the idea of "balanceOrUpdateMetrics", I will resubmit a new commit based on this idea.